### ISSUE TITLE
Enable locally runnable visualiser

### ISSUE THREAD

**Issue by zanete** [Green-Software-Foundation/if#982]

**Comment by zanete** [comment]
hey @osamajandali I hope you have had a chance to check out the video that @jawache sent on the 9th of August about the IF website. 

He mentioned needing to film a quick walk through where he wants to be able to frequently visualise a manifest file that's only available on the local machine without needing to upload it. He envisages running `if-vis` on the local machine command line which would somehow pull up a browser with the visualiser running on the manifest file that was passed as an argument. 

I wanted to hear your thoughts on how this could work. Please give your opinion on how this could be implemented and what development steps would need to be done. 🙏 

cc @jawache @jmcook1186

**Comment by osamajandali** [comment]
I agree with the approach of creating a command that takes the manifest file and sends it to the visualizer. The next step is to determine the mechanism for sending the file. Given the file size, it cannot be stringified and sent as a parameter via the URL.

I suggest that when the command is executed, it should upload the file to a CDN (e.g., S3). The visualizer can then fetch the file from the CDN URL and proceed to handle it as it normally would.

**Comment by jawache** [comment]
Thanks @zanete and @osamajandali  generally I'm against uploading it to something like S3 since people might have sensitive data on their manifest files.

For the demo can I just run the visualizer locally and then drag a local filesystem file in there?

I was proposing in the video maybe we can just POST the file over https and never store it?

Best solution is a command that just opens a localhost browser with visualizer running and the file passed via localhost or something?

Basically, best solution is a solution where the manifest file doesn't leave your computer.

**Comment by zanete** [comment]
fyi @jawache @osamajandali, @narekhovhannisyan will do some investigation on how the manifest could be passed to the visualiser and will let us know

**Comment by jawache** [comment]
@narekhovhannisyan and folk can you describe in detail what is being proposed as a solution?

**Comment by narekhovhannisyan** [comment]
@jawache @zanete @osamajandali if we are going to run `if-vis` in the command line and pass the file path to it, then the solution can be: 

1. Downloading/reading file.
2. Passing that file to the server before starting, then it will render the result by injecting it into template (in this case website should be implemented using templating f.ex. [pug](https://pugjs.org/api/getting-started.html) or [ejs](https://ejs.co/)). The server will render it using the given yaml.

**Comment by narekhovhannisyan** [comment]
Locally bootstrapped something like that.
<img width="1512" alt="Screenshot 2024-08-28 at 12 41 29" src="https://github.com/user-attachments/assets/451b0223-b532-4f7a-8b54-18cebb1f582c">

**Comment by jawache** [comment]
Gotcha, so just POST the local file to a local (?) web server. On POST web server injects the YAML in the JavaScript as a string and then when the visualizer loads, the YAML is right there for it to load and display?

If so that works for me and the demo onstage, will look good.

if-vis doesn't need to launch a browser, I will just open one up pointed to localhost:8080, make it fit the right size of the screen, and just refresh that. It will be annoying if it keeps on opening up a browser every time I run it, and then I have to make sure that new browser window is in the right place.

If the above is the approach let's just use localhost for now.

In fact, I'd say if-vis IS the local web server, when you run if-vis, it just launches a localhost:8080 server which if you load has the YAML injected into it!

**Comment by narekhovhannisyan** [comment]
@jawache almost yeah, I think for not making things complicated with requests, we can just await for the readFile operation, and then render the website with the YAML injected. WDYT?

**Comment by zanete** [comment]
Final resolution:
- the visualiser will be a react app which will be reused in the hosted as well as the local version
- @osamajandali and @narekhovhannisyan to discuss the implementation closer the time, but roughly the local will be run on command line as `if-vis` with the manifest path, the bash script will start the server that serves the react frontend.
